Make the C API higher level

ferdonline opened the pull request:
This PR builds on #2357 and makes the API more user friendly, namely easier to use and harder to make mistakes.

Its most important property is the stack protection:
 - functions' argument count is determined from the argument types
 - functions' return value is automatically popped to a result object

It was lightly inspired by Python C API, namely function/method calls expect all parameters as var-args.

## Scope
 - Definition of an argument passing format
 - Make function call use the format and variable number of args
 - Do error handling

## Tests
 - Unit tests adapted accordingly. Please refer to them to see user facing impact.

ferdonline committed:
A higher level API for function calls

ferdonline committed:
Working version of the High Level API

ferdonline committed:
Cleanup

ferdonline committed:
Documentation and more cleanup

1uc commented on test/api/vclamp.cpp:
```diff
@@ -58,33 +58,26 @@ int main(void) {
-    v = nrn_object_new(nrn_symbol("Vector"), 0);
-    nrn_rangevar_push(nrn_symbol("v"), soma, 0.5);
-    nrn_double_push(1);
-    nrn_method_call(v, nrn_method_symbol(v, "record"), 2);
```
I think this is a case of pushing a double and a range variable without popping them again, i.e. and example of something this PR improves.

1uc commented on test/api/vclamp.cpp:
```diff
@@ -58,33 +58,26 @@ int main(void) {
 
     // voltage clamp at soma(0.5)
-    nrn_double_push(0.5);
-    vclamp = nrn_object_new(nrn_symbol("VClamp"), 1);
```
Looks like a push without a pop.

1uc commented on test/api/vclamp.cpp:
```diff
@@ -39,13 +36,16 @@ int main(void) {
-    temp_str = strdup("stdrun.hoc");
-    nrn_str_push(&temp_str);
-    nrn_function_call(nrn_symbol("load_file"), 1);
-    nrn_double_pop();
```
Looks like pushing a string and popping a double.

1uc commented on test/api/netcon.cpp:
```diff
@@ -55,50 +47,43 @@ int main(void) {
-    v = nrn_object_new(nrn_symbol("Vector"), 0);
-    nrn_rangevar_push(nrn_symbol("v"), soma, 0.5);
-    nrn_double_push(20);
-    nrn_method_call(v, nrn_method_symbol(v, "record"), 2);
```
Looks like pushing without popping.

ferdonline replied:
With the next API we hope to avoid any manual handling of the stack. But what was there looked ok since the function would pop from the stack the number of arguments as given, in this case 2.

ferdonline committed:
No more manual stack pop. Methods pop the right type and return `Result`

ferdonline committed:
Creating catch unit tests for API

ferdonline committed:
Make return value pop its own func. Investigate other internal funcs

By default Neuron functions return double. However, some internal
functions can retuen also Objects or Strings (STRINGFUNC and OBJECTFUNC)
Unfortuinately those cases are not supported directly and for the moment
we issue a warning.

ferdonline commented:
@pramodk @1uc Is there still any interest in this one? I feel like the original API was a bit cumbersome and it was interesting to offer something more intuitive. But if we wait too much we risk the initial one starts being used and then we're locked

ferdonline committed:
Merge fixes

ferdonline commented:
I have rebased the branch, which was a major obstacle to review and understand the real impact.

ferdonline committed:
Replace new R123 noise call

ferdonline committed:
Drop weak mod_reg, it wont work for corenrn on some compilers

ramcdougal commented on test/api/test_common.h:
```diff
@@ -1,29 +1,52 @@
+
+// --- String Helpers (Neuron doesnt handle char* strings ? yet)
+
+inline NrnResult nrn_function_call_s(const char* func_name, const char* v) {
```
A function like this doesn't eliminate the need for pushing a string to the stack... e.g., we still need that for a function that takes a string as a second argument, etc

ramcdougal replied:
If we really wanted to provide a generic way of calling things, we could have a function that takes a list of types (roughly analagous to `Py_BuildValue` or `printf`) and then the variable length list... but it wouldn't be just for function calling; it would have to be here, in the constructors, and who knows where else.

One other consideration: I don't know how well generic external function interface libraries (e.g., `ctypes`) play with variable length argument lists of various types. (Normally, you can give `ctypes` the prototype and it'll transparently do the necessary conversion from e.g., a Python string to a `char*`.)

ferdonline replied:
This function is a helper for the tests (note it's part of `test_common.h`). I think there's some value in keeping examples clean. People can always expand and do what the function does internally in their code.

ramcdougal commented on test/api/vclamp.cpp:
```diff
@@ -32,11 +32,17 @@ int main(void) {
+        std::cerr << "Err: " << nrn_stack_err() << std::endl;
+    }
+    // Do it right this time
+    nrn_function_call_s("load_file", "stdrun.hoc");
```
`nrn_function_call_s` only works for one specific prototype. I don't think we want anything that specific in our API or in our examples.

ferdonline replied:
See comment above

ramcdougal commented on src/nrniv/neuronapi.cpp:
```diff
@@ -158,159 +168,277 @@ void nrn_segment_diam_set(Section* const sec, const double x, const double diam)
 
-void nrn_rangevar_push(Symbol* sym, Section* sec, double x) {
-    hoc_push(nrn_rangepointer(sec, sym, x));
+RangeVar nrn_rangevar_new(Section* sec, double x, const char* var_name) {
```
No idea if this would have worked before, but if we're embedding the `node` inside the `RangeVar` object, wouldn't that mean we might be asking for trouble if `nseg` changes?

ferdonline replied:
I am not sure I can answer that. True that this code assumes RangeVar is tied to a node/segment. If that must not happen we should change this for do the lookup on the very last moment @nrnhines

ramcdougal commented on src/nrniv/neuronapi.cpp:
```diff
@@ -158,159 +168,277 @@ void nrn_segment_diam_set(Section* const sec, const double x, const double diam)
-void nrn_double_push(double val) {
-    hoc_pushx(val);
+static int stack_push_args(const char arg_types[], va_list args) {
+    for (int n_arg = 0; n_arg < 32; ++n_arg) {  // max 32 args
```
why a limit of 32? Is that intrinsic to HOC as configured? If so, can we have both this and that use the same constant instead of a magic number here?

ferdonline replied:
If there's a related hoc limit defined in a constant I'd be happy to use.

ferdonline commented:
@ramcdougal Thanks for the remarks. I have given some feedback.
Also, can you indicate where I could refer to the functions which are currently used and must not be changed? We currently don't really have tests and would be great we start a suite to enforce we don't break required API.
